tools/ci_connector_ops/ci_connector_ops/pipelines/publish.py

```diff
@@ -143,55 +151,55 @@ async def run_connector_publish_pipeline(context: PublishConnectorContext, semap
     1. Validate the metadata file.
     2. Check if the connector image already exists.
     3. Build the connector, with platform variants.
-    4. Upload its spec to the spec cache bucket.
-    5. Push the connector to DockerHub, with platform variants.
+    4. Push the connector to DockerHub, with platform variants.
+    5. Upload its spec to the spec cache bucket.
     6. Upload its metadata file to the metadata service bucket.
 
     Returns:
         ConnectorReport: The reports holding publish results.
     """
-    async with semaphore:
-        async with context:
 
-            metadata_upload_step = metadata.MetadataUpload(context)
-            steps_before_ready_to_publish = [
-                metadata.MetadataValidation(context, context.metadata_path),
-                CheckConnectorImageDoesNotExist(context),
-            ]
-            steps_before_ready_to_publish_results = await run_steps(steps_before_ready_to_publish)
+    def create_connector_report(results: List[StepResult]) -> ConnectorReport:
+        report = ConnectorReport(context, results, name="PUBLISH RESULTS")
+        context.report = report
+        return report
 
-            if steps_before_ready_to_publish_results[-1].status is not StepStatus.SUCCESS:
-                if steps_before_ready_to_publish_results[-1].status is StepStatus.SKIPPED:
+    async with semaphore:
+        async with context:
+            results = []
+            metadata_validation_results = await metadata.MetadataValidation(context, context.metadata_path).run()
+            results.append(metadata_validation_results)
+            if metadata_validation_results.status is not StepStatus.SUCCESS:
+                return create_connector_report(results)
+
+            check_connector_image_results = await CheckConnectorImageDoesNotExist(context).run()
+            results.append(check_connector_image_results)
+            if check_connector_image_results.status is not StepStatus.SUCCESS:
+                if check_connector_image_results.status is StepStatus.SKIPPED:
                     context.logger.info(
                         "The connector version is already published. Let's upload metadata.yaml to GCS even if no version bump happened."
                     )
```

Comment:
nit: We could remove one layer of nesting if we breakout the `on skip` logic

```suggestion
            check_connector_image_results = await CheckConnectorImageDoesNotExist(context).run()
            results.append(check_connector_image_results)
            
            # Run metadata upload if skipped
            if check_connector_image_results.status is StepStatus.SKIPPED:            
                context.logger.info(
                        "The connector version is already published. Let's upload metadata.yaml to GCS even if no version bump happened."
                )
                metadata_upload_results = await metadata.MetadataUpload(context).run()
                results.append(metadata_upload_results)
                
            # Exit if not success
            if check_connector_image_results.status is not StepStatus.SUCCESS:
                return create_connector_report(results)
```